ebean-core/src/main/java/io/ebeaninternal/server/type/ScalarTypeJsonObjectMapper.java

```diff
@@ -163,11 +182,52 @@ public T read(DataReader reader) throws SQLException {
         return null;
       }
       try {
-        return objectReader.readValue(json, deserType);
+        T ret = objectReader.readValue(json, deserType);
+        track(ret, json);
+        return ret;
       } catch (IOException e) {
         throw new TextException("Failed to parse JSON [{}] as " + deserType, json, e);
       }
     }
+    
+
+    private Map<TrackReference<T>, String> originalJson = new ConcurrentHashMap<>();
+    private final ReferenceQueue<Object> queue = new ReferenceQueue<>();
+    
+    private static class TrackReference<T> extends WeakReference<T> {
+
+      private final int hashCode;
+
+      public TrackReference(T referent) {
+        super(referent);
+        hashCode = System.identityHashCode(referent);
+      }
+      
+      public TrackReference(T referent, ReferenceQueue<? super T> q) {
+        super(referent, q);
+        hashCode = System.identityHashCode(referent);
+      }
+      
+      @Override
+      public int hashCode() {
+        return hashCode;
+      }
+      @Override
+      public boolean equals(Object other) {
+        return other instanceof TrackReference && this.get() == ((TrackReference<?>)other).get();
+      }
+    }
+    
+    private void track(T obj, String json) {
+      if (obj != null) {
+        TrackReference<T> ref = new TrackReference<T>(obj, queue);
+        originalJson.put(ref, json);
```

Comment:
I do not use a WeakHashMap here, as the map uses equals/hashcode of the json object and I would not find the objects again in the map.... (While writing this line, I think this is no problem)
TODO: check if weakHashMap can work as well here...

Reply:
WeakHashMap will not work here, if the same object is deserialized several times, the following may be true:
`object1.equals(object2)`  and `object1 != object2` 
if `object1` is gc'ed then you may not find `object2` in the map any more.